### PR TITLE
KAFKA-9113: Fix system / unit tests

### PR DESCRIPTION
**Pull request by guozhangwang**
1. Address comments from previous PRs.
2. Rebase from trunk (primarily KIP-562).
3. Fix system tests.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by guozhangwang** [comment]
Triggered three for previously failed ones: 
https://jenkins.confluent.io/job/system-test-kafka-branch-builder/3721/
https://jenkins.confluent.io/job/system-test-kafka-branch-builder/3722/
https://jenkins.confluent.io/job/system-test-kafka-branch-builder/3723/

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/AbstractTask.java [hunk 1]
```diff
@@ -16,11 +16,55 @@
...
+                 final ProcessorStateManager stateMgr,
+                 final Set<TopicPartition> partitions) {
+        this.id = id;
+        this.stateMgr = stateMgr;
+        this.topology = topology;
+        this.partitions = partitions;
+        this.stateDirectory = stateDirectory;
+    }
+
+    @Override
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/AbstractTask.java:48]: Move common functions here @mjsax.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreChangelogReader.java [hunk 2]
```diff
@@ -188,6 +188,10 @@ int bufferedLimitIndex() {
     // to update offset limit for standby tasks;
     private Consumer<byte[], byte[]> mainConsumer;
 
+    // the flag indicating limit offsets could be updated --- this is only needed for standby tasks that have limit
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreChangelogReader.java:191]: This is to clear one TODO to update limit offsets.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java [hunk 3]
```diff
@@ -238,7 +228,7 @@ public void completeRestoration() {
      */
     @Override
     public void suspend() {
-        if (state() == State.CLOSING || state() == State.SUSPENDED) {
+        if (state() == State.CREATED || state() == State.CLOSING || state() == State.SUSPENDED) {
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:231]: This is found that when we are in created state (see below) we should not throw illegal-state exception.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/Task.java [hunk 4]
```diff
@@ -66,42 +67,12 @@
...
     enum State {
-        CREATED(1, 4) {         // 0
-            @Override
-            public boolean hasBeenRunning() {
-                return false;
-            }
-        },
-        RESTORING(2, 3, 4) {    // 1
-            @Override
-            public boolean hasBeenRunning() {
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/Task.java:77]: As suggested by @mjsax we can now remove this func.

File: streams/src/test/java/org/apache/kafka/streams/integration/StoreQueryIntegrationTest.java [hunk 5]
```diff
@@ -239,15 +239,18 @@ public void shouldQueryAllStalePartitionStores() throws Exception {
...
-        final ReadOnlyKeyValueStore<Integer, Integer> store1 = kafkaStreams1
-                .store(StoreQueryParams.fromNameAndType(TABLE_NAME, queryableStoreType).enableStaleStores());
-
-        final ReadOnlyKeyValueStore<Integer, Integer> store2 = kafkaStreams2
-                .store(StoreQueryParams.fromNameAndType(TABLE_NAME, queryableStoreType).enableStaleStores());
 
         // Assert that both active and standby are able to query for a key
-        assertThat(store1.get(key), is(notNullValue()));
-        assertThat(store2.get(key), is(notNullValue()));
+        TestUtils.waitForCondition(() -> {
```
Comment (guozhangwang) [inline streams/src/test/java/org/apache/kafka/streams/integration/StoreQueryIntegrationTest.java:244]: This is to fix a flaky test I found while working on the PR: even on trunk it is flaky too. The reason is that standby tasks may not fully restore even if it has transited to running.

File: tests/kafkatest/tests/streams/streams_broker_compatibility_test.py [hunk 6]
```diff
@@ -79,8 +79,6 @@ def test_fail_fast_on_incompatible_brokers_if_eos_enabled(self, broker_version):
     @parametrize(broker_version=str(LATEST_1_1))
     @parametrize(broker_version=str(LATEST_1_0))
     @parametrize(broker_version=str(LATEST_0_11_0))
-    @parametrize(broker_version=str(LATEST_0_10_2))
```
Comment (guozhangwang) [inline tests/kafkatest/tests/streams/streams_broker_compatibility_test.py:82]: @ableegoldman this is to fix the broken system tests from the disable topic creation PR.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java [hunk 7]
```diff
@@ -185,11 +187,6 @@ public void handleAssignment(final Map<TaskId, Set<TopicPartition>> activeTasks,
             logPrefix
         );
 
-        // initialize the created tasks
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:188]: This is a found regression: initializeIfNeeded may fail to lock state dir if there are multiple threads, in which case we should just skip the initialization process and retry in the next loop.

**Comment by guozhangwang** [comment]
I'm merging this back to `base` now in order to trigger another unit / system directly, please feel free to continue reviewing and I will address comments in further PRs.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java [hunk 8]
```diff
@@ -209,15 +206,29 @@ private void addNewTask(final Task task) {
...
+            }
+
             if (task.state() == RESTORING) {
                 restoringTasks.add(task);
             }
         }
 
-        boolean allRunning = true;
-        if (!restoringTasks.isEmpty()) {
+        if (allRunning && !restoringTasks.isEmpty()) {
```
Comment (vvcephei) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:231]: Not sure this conditional makes sense... It says "if all the tasks are running, and there are restoring tasks", which is self-contradictory. Maybe we can have two boolean flags, `all initialized` and `all running`?

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java [hunk 9]
```diff
@@ -236,15 +235,23 @@ public void shouldAddNewActiveTasks() {
...
+        replay(consumer, activeTaskCreator, standbyTaskCreator, changeLogReader);
 
         taskManager.handleAssignment(assignment, emptyMap());
 
-        assertThat(task00.state(), is(Task.State.RESTORING));
+        assertThat(task00.state(), is(Task.State.CREATED));
+
+        taskManager.checkForCompletedRestoration();
+
+        assertThat(task00.state(), is(Task.State.RUNNING));
```
Comment (vvcephei) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:254]: This seems out of the scope of this test, which is just that the new task gets created. It's probably redundant with another test now.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java (hunk 8 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:231]: The rationale is that some tasks may still be in CREATED state and some maybe in RESTORING, in either case we should set `allRunning` to false. If some are still in CREATED state then `allRunning` would be false here already so the first condition is just to save the second check.

So I think correctness-wise that is fine but maybe a bit confusing, I can make this logic a bit cleaner now.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java (hunk 9 above)
Comment (guozhangwang) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:254]: Yeah I can do that.

**Pull request merged**